File: linked-list/2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(0)
        cur = dummy
        p, q = l1, l2

        k = 0

        while p or q:
            x1 = p.val if p else 0
            x2 = q.val if q else 0

            sum = x1 + x2 + k

            k = sum // 10
            sum = sum % 10

            cur.next = ListNode(sum)
            cur = cur.next

            if p:
                p = p.next
            if q:
                q = q.next

        if k > 0:
            cur.next = ListNode(k)

        return dummy.next
# @lc code=end

File: Golang/2.add-two-numbers.go
/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 */
package Golang

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{Val: 0}
	cur := dummy

	curry := 0
	a, b := 0, 0

	for l1 != nil || l2 != nil || curry != 0 {
		if l1 != nil {
			a = l1.Val
			l1 = l1.Next
		} else {
			a = 0
		}

		if l2 != nil {
			b = l2.Val
			l2 = l2.Next
		} else {
			b = 0
		}

		sum := a + b + curry
		num := sum % 10
		curry = sum / 10

		cur.Next = &ListNode{Val: num}
		cur = cur.Next
	}

	return dummy.Next
}

// @lc code=end
